Add support for --package flag

THenry14 opened the pull request:
-- * Extract build from declare (https://github.com/foundry-rs/starknet-foundry/pull/1565)
-> * Add support for --package flag
-- * Add workspace tests (https://github.com/foundry-rs/starknet-foundry/pull/1572)

todo:
-- * cleanup scarb_utils, use things from scarb_api/scarb_ui

## Introduced changes

Closes https://github.com/foundry-rs/starknet-foundry/issues/17
Closes https://github.com/foundry-rs/starknet-foundry/issues/449

<!-- A brief description of the changes -->

- Adds support for `--package` flag that can be used to determine the scarb package to work with

## Checklist

<!-- Make sure all of these are complete -->

- [X] Linked relevant issue
- [X] Updated relevant documentation
- [X] Added relevant tests
- [X] Performed self-review of the code
- [X] Added changes to `CHANGELOG.md`

MaksymilianDemitraszek commented on CHANGELOG.md:
```diff
@@ -21,6 +21,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+#### Added
+
+- `--package` flag to specify scarb package to work with
```
To which command? What does it mean "to work with"?

THenry14 replied:
Hmm, perhaps you're right. I have added it as global for now, since other things use it indirectly (basically everything that touches cast config, eg accounts). This was because I thought that peraps packages should be able to have their own configs, but I'm not sure now. wydt?

MaksymilianDemitraszek replied:
I think we shouldn't include it when command does not operate on packages. 

Commands having their own configs is not a problem of foundry, scarb should resolve it based on where command is run from.

THenry14 replied:
Wait, I'm talking about cast config ([tools.sncast.proflie])

THenry14 replied:
Ok, to move things forward I'll remove such possibility (to define configs per package), as perhaps it indeed does not make much sense for some commands. One, top-level config per workspace should be en ough, especially so that we allow for multiple profiles

MaksymilianDemitraszek replied:
Agreed this should be investigated, we can look into later and discuss it with scarb team.

MaksymilianDemitraszek replied:
Can we create an issue for investigating how per package configs work?

piotmag769 replied:
Wdym by top-level config?

THenry14 replied:
yes, this is indeed an issue. So when we have per subcommands --package flag, we'll probably need to check members first, take every tool.sncast config from those packages and then check for default config (without profile name) - at least that's the only sensible solution I can think of. Main drawback of this is one can only have one default (tool.scast) profile per workspace. thoughts?

THenry14 replied:
Discussed with @piotmag769 possible solutions to this pickle, and we came up with this flow:

```
- we try to declare/script
    - for workspaces:
        - if there is only one package -> return its castconfig (if it exists, else empty)
        - more than one package -> require `--package` flag, return its castconfig (if it exists, else empty)
    - for single packages:
        - return its castconfig (if it exists, else empty)

- we try to invoke/call/account/whatever else:
    - for workspaces:
        - if there is only one package -> return its castconfig (if it exists, else empty)
        - more than one package -> print out warning that package couldn not be determined, and return an empty castconfig
    - for single packages:
        - return its castconfig (if it exists, else empty)
```

It's not ideal, but allows for using configs from packages for subcommands supporting it, and makes other cases more explicit.

**An alternative approach** would be to revert the `--package only to declare/script` altogether, and make it global again which would allow to pick workspaces (and thus, ccastconfigs) for each subcommand. That would be a cleaner and easier solution, but would mean that (in some cases) we must specify a package to invoke some random contract, which is far from perfect. 

Thoughts? Is the solution 1) perfectly acceptable, should we lean to solution 2) with the alternative approach, or maybe something different altogether?

MaksymilianDemitraszek commented on crates/sncast/src/helpers/scarb_utils.rs:
```diff
@@ -196,48 +197,63 @@ pub fn get_scarb_metadata_with_deps(
-pub fn verify_or_determine_scarb_manifest_path(
-    path_to_scarb_toml: &Option<Utf8PathBuf>,
-) -> Option<Utf8PathBuf> {
+pub fn ensure_scarb_manifest_path(path_to_scarb_toml: &Option<Utf8PathBuf>) -> Result<Utf8PathBuf> {
```
I don't understand what this function does from the name.

THenry14 replied:
Carry over from https://github.com/foundry-rs/starknet-foundry/pull/1565#discussion_r1464663169

This function ensures valid path to manifest file. I agree that name may be a bit unfortunate, maybe I should change it to something like `get_scarb_manifest`?

MaksymilianDemitraszek replied:
If it returns scarb manifest i think it would be a better name.

THenry14 replied:
it turned out there is a funtion named `get_scarb_manifest` (it uses scarb to get the manifest), so I have renamed `ensure_scarb_manifest_path` to `assert_manifest_path_exists`, as after some changes this function actually asserts its existence. (https://github.com/foundry-rs/starknet-foundry/pull/1569/commits/bcc6b0fac2cb70a8deddb51e6ba380034d1574d2)

THenry14 replied:
Updated docs + changelog (please check the docs especially, as I'm afraid I didn't stress enough the fact that only most top-level Scarb.toml holds the config https://github.com/foundry-rs/starknet-foundry/pull/1569/commits/40b52d3fb1ea6c6a83570824c676f582aa1d383e)

Moved flags to respective subcommands https://github.com/foundry-rs/starknet-foundry/pull/1569/commits/4791a4d94c28f71cd8e82dc90c9a358a283d8581

THenry14 commented on docs/src/appendix/sncast/common.md:
```diff
@@ -12,6 +12,13 @@ Path to `Scarb.toml` file.
+
+Name of the package that should be used.
+
+If supplied, this package will be used for all operations. Required if more than one package exists in a workspace.
```
For now - yes in my opinion. We don't have cairo paths in scarb artifacts as of now, so specifying a package is our only choice to pick specific module if we're in a workspace with multiple packages.

MaksymilianDemitraszek commented:
So the final conclusion is we make `--package` a per command flag, yes?

THenry14 commented:
> So the final conclusion is we make `--package` a per command flag, yes?

Yes, that is correct. I'm working on the changes, but had to jump in to some reviews etc. Will commit later today.

MaksymilianDemitraszek reviewed (commented):
Submitting so you can re-request when its ready.

THenry14 commented:
> > So the final conclusion is we make `--package` a per command flag, yes?
> 
> Yes, that is correct. I'm working on the changes, but had to jump in to some reviews etc. Will commit later today.

Updated docs + changelog (please check the docs especially, as I'm afraid I didn't stress enough the fact that only most top-level Scarb.toml holds the config https://github.com/foundry-rs/starknet-foundry/pull/1569/commits/40b52d3fb1ea6c6a83570824c676f582aa1d383e)

Moved flags to respective subcommands https://github.com/foundry-rs/starknet-foundry/pull/1569/commits/4791a4d94c28f71cd8e82dc90c9a358a283d8581

piotmag769 commented on crates/sncast/src/helpers/scarb_utils.rs:
```diff
@@ -191,43 +192,69 @@ pub fn get_scarb_metadata_with_deps(
+    let main_package = metadata
+        .packages
+        .into_iter()
+        .find(|package| package.manifest_path == metadata.workspace.manifest_path)
+        .expect("Failed to find main package in metadata");
```
This will fail for virtual workspaces I believe
https://docs.swmansion.com/scarb/docs/reference/workspaces.html#virtual-workspace

THenry14 replied:
yup, you are right. I have rebased https://github.com/foundry-rs/starknet-foundry/pull/1572 on friday but did not checked its status, and its failing. trying to figure out a fix now.

THenry14 commented on crates/sncast/src/main.rs:
```diff
@@ -149,6 +151,9 @@ fn main() -> Result<()> {
         print_command_result("script", &mut result, numbers_format, &output_format)?;
         Ok(())
     } else {
+        let mut config = load_config(&cli.profile, &None)?;
```
improving this will be a part of https://github.com/foundry-rs/starknet-foundry/issues/1329

piotmag769 commented on crates/sncast/src/helpers/scarb_utils.rs:
```diff
@@ -63,43 +63,59 @@ pub fn get_scarb_metadata_with_deps(manifest_path: &Utf8PathBuf) -> Result<Metad
 }
 
-pub fn get_package_metadata<'a>(
+fn get_package_metadata_by_name<'a>(
```
This is prob `get_workspace_member_metadata_by_name`, right? It should use `metadata.workspace. members`

THenry14 replied:
yup, as we discussed offline - this will be a part of scarb_utils cleanup mentioned in pr description :+1:

abulenok commented on crates/sncast/src/helpers/scarb_utils.rs:
```diff
@@ -63,43 +63,59 @@ pub fn get_scarb_metadata_with_deps(manifest_path: &Utf8PathBuf) -> Result<Metad
+fn get_default_package_metadata(metadata: &Metadata) -> Result<&PackageMetadata> {
+    match metadata.packages.iter().collect::<Vec<_>>().as_slice() {
+        [package] => Ok(package),
+        [] => Err(anyhow!("No package found in metadata")),
+        _ => Err(anyhow!(
+            "More than one package found in metadata - specify package using --package flag"
```
```suggestion
        [] => Err(anyhow!("No package found in scarb metadata")),
        _ => Err(anyhow!(
            "More than one package found in scarb metadata - specify package using --package flag"
```

abulenok commented on crates/sncast/src/starknet_commands/declare.rs:
```diff
@@ -45,7 +49,7 @@ pub async fn declare(
     let contract_artifacts = artifacts
         .get(&contract_name)
-        .ok_or(anyhow!("Failed to find artifacts in starknet_artifacts.json file. Please ensure you have enabled sierra and casm code generation in Scarb.toml"))?;
+        .ok_or(anyhow!("Failed to find artifacts in starknet_artifacts.json file. Please make sure you have specified correct package using `--package` flag and that you have enabled sierra and casm code generation in Scarb.toml."))?;
```
It would be nice to include the contract name in the error message.

Smth like `Failed to find artifacts in starknet_artifacts.json file for {contract_name} contract`

THenry14 replied:
the diff showed, because I merged a few PRs and did not rebase - I have now done this, and contract name is included in the message.
